README.md
# convolution

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![npm download][download-image]][download-url]

Convolution using the FFT or direct algorithm.

## Installation

```console
npm install ml-convolution
```

## Usage

### One execution

```js
import { directConvolution, fftConvolution } from 'ml-convolution';

const input = [0, 1, 2, 3];
const kernel = [-1, 1, -1];

const outputDirect = directConvolution(input, kernel); // [-1, -1, -2, 1]
const outputFFT = fftConvolution(input, kernel); // [-1, -1, -2, 1]
```

The functions both take an optional third argument to determine the way borders
are processed. The default value, `CONSTANT`, will consider that the values out
of the bounds are all 0. If it is set to `CUT`, borders will be ignored and the
result will be smaller than te input by `kernel.length - 1`:

```js
const outputDirect = directConvolution(input, kernel, 'CUT'); // [-1, -2]
```

### Optimized, multiple executions

If you need to execute the convolution many times with the same kernel and input
length, you should consider instead to use the class-based API:

```js
import { DirectConvolution, FFTConvolution } from 'ml-convolution';

// const input = [0, 255, 255, 255, 255, 0, 0, 0];
const kernel = [0.1, 0.2, 0.3];

// First parameter is the size of the inputs and allows to pre-allocate an array with the correct size
const direct = new DirectConvolution(8, kernel, 'CUT');

// The convolve function mutates the same array at each execution
direct.convolve([0, 255, 255, 255, 255, 0, 0, 0]); // [ 127.5, 153, 153, 76.5, 25.5, 0 ]
direct.convolve([255, 0, 0, 255, 255, 255, 0, 0]); // [ 25.5, 76.5, 127.5, 153, 76.5, 25.5 ]

const fft = new FFTConvolution(8, kernel, 'CONSTANT');
fft.convolve([0, 255, 255, 255, 255, 0, 0, 0]); // [ 76.5, 127.5, 153, 153, 76.5, 25.5, 0, 0 ]
fft.convolve([255, 0, 0, 255, 255, 255, 0, 0]); // [ 51, 25.5, 76.5, 127.5, 153, 76.5, 25.5, 0 ]
```

### Benchmark

With small kernels, direct convolution is usually faster:  
Current results suggest that from a kernel size around 64, FFT convolution should be used.

| Data x Kernel | fft [ops/s] | direct [ops/s] |
| ------------- | ----------- | -------------- |
| 128 x 5       | 97889       | 569110         |
| 128 x 11      | 99403       | 280271         |
| 128 x 17      | 97686       | 181608         |
| 128 x 33      | 94633       | 93847          |
| 128 x 65      | 96585       | 49320          |
| 128 x 129     | 97189       | 25346          |
| 128 x 513     | 21771       | 6469           |
| 512 x 5       | 20712       | 144025         |
| 512 x 11      | 21134       | 73189          |
| 512 x 17      | 21201       | 44320          |
| 512 x 33      | 21037       | 23591          |
| 512 x 65      | 21398       | 12405          |
| 512 x 129     | 21514       | 6358           |
| 512 x 513     | 21494       | 1618           |
| 2048 x 5      | 4746        | 36360          |
| 2048 x 11     | 4740        | 18422          |
| 2048 x 17     | 4735        | 11248          |
| 2048 x 33     | 4689        | 5927           |
| 2048 x 65     | 4740        | 3100           |
| 2048 x 129    | 4741        | 1591           |
| 2048 x 513    | 4753        | 405            |
| 4096 x 5      | 2068        | 18201          |
| 4096 x 11     | 2062        | 9241           |
| 4096 x 17     | 2071        | 5629           |
| 4096 x 33     | 2069        | 2976           |
| 4096 x 65     | 2079        | 1551           |
| 4096 x 129    | 2074        | 797            |
| 4096 x 513    | 2079        | 203            |
| 16384 x 5     | 370         | 4036           |
| 16384 x 11    | 371         | 2295           |
| 16384 x 17    | 377         | 1390           |
| 16384 x 33    | 374         | 748            |
| 16384 x 65    | 370         | 389            |
| 16384 x 129   | 375         | 199            |
| 16384 x 513   | 376         | 51             |
| 65536 x 5     | 70          | 991            |
| 65536 x 11    | 70          | 541            |
| 65536 x 17    | 70          | 351            |
| 65536 x 33    | 69          | 186            |
| 65536 x 65    | 71          | 97             |
| 65536 x 129   | 71          | 50             |
| 65536 x 513   | 70          | 13             |
| 262144 x 5    | 10          | 247            |
| 262144 x 11   | 10          | 135            |
| 262144 x 17   | 10          | 88             |
| 262144 x 33   | 10          | 47             |
| 262144 x 65   | 10          | 24             |
| 262144 x 129  | 10          | 12             |
| 262144 x 513  | 10          | 3              |
| 1048576 x 5   | 2           | 60             |
| 1048576 x 11  | 2           | 32             |
| 1048576 x 17  | 2           | 22             |
| 1048576 x 33  | 2           | 12             |
| 1048576 x 65  | 2           | 6              |
| 1048576 x 129 | 2           | 3              |
| 1048576 x 513 | 2           | 1              |

## License

[MIT](./LICENSE)

[npm-image]: https://img.shields.io/npm/v/ml-convolution.svg?style=flat-square
[npm-url]: https://npmjs.org/package/ml-convolution
[travis-image]: https://img.shields.io/travis/mljs/convolution/master.svg?style=flat-square
[travis-url]: https://travis-ci.org/mljs/convolution
[download-image]: https://img.shields.io/npm/dm/ml-convolution.svg?style=flat-square
[download-url]: https://npmjs.org/package/ml-convolution


History.md
# [2.0.0](https://github.com/mljs/convolution/compare/v1.0.2...v2.0.0) (2019-04-15)


### Bug Fixes

* **fft:** make FFT convolution give a correct result ([e3fe4b7](https://github.com/mljs/convolution/commit/e3fe4b7))


### Features

* add FFTConvolution class ([4700d8b](https://github.com/mljs/convolution/commit/4700d8b))


### BREAKING CHANGES

* **fft:** fftConvolution no also expects the borderType option.



<a name="0.2.0"></a>
# [0.2.0](https://github.com/mljs/convolution/compare/v0.1.0...v0.2.0) (2017-05-22)


### Features

* **directConvolution:** add optional output argument ([6e06a5a](https://github.com/mljs/convolution/commit/6e06a5a))



<a name="0.1.0"></a>
# 0.1.0 (2017-05-22)





src/__tests__/test.js
import convolution from '..';

describe('convolution', () => {
    it('with one value kernel', () => {
        expect(convolution([0, 1, 2, 3], [0])).toEqual([0, 0, 0, 0]);
        expect(convolution([0, 1, 2, 3], [1])).toEqual([0, 1, 2, 3]);
        expect(convolution([0, 1, 2, 3], [2])).toEqual([0, 2, 4, 6]);
    });

    it('with three values kernel', () => {
        expect(convolution([0, 1, 2, 3], [1, 1, 1])).toEqual([0, 1, 3, 6, 5, 3]);
        expect(convolution([0, 1, 2, 3], [-1, 1, -1])).toEqual([0, -1, -1, -2, 1, -3]);
    });
});


src/__tests__/fftConvolution.js
import { fftConvolution, FFTConvolution } from '..';

/* eslint jest/expect-expect: ["error", { "assertFunctionNames": ["expect", "checkClose"] }] */

describe('fft convolution', () => {
  it('fft with one value kernel', () => {
    expect(fftConvolution([0, 1, 2, 3], [0])).toStrictEqual([0, 0, 0, 0]);
    expect(fftConvolution([0, 1, 2, 3], [1])).toStrictEqual([0, 1, 2, 3]);
    expect(fftConvolution([0, 1, 2, 3], [2])).toStrictEqual([0, 2, 4, 6]);
  });

  it('fft with three values kernel', () => {
    checkClose(fftConvolution([0, 1, 2, 3], [1, 1, 1]), [1, 3, 6, 5]);
    // prettier-ignore
    checkClose(fftConvolution([0, 1, 2, 3], [-1, 1, -1]), [-1, -1, -2, 1]);
  });

  it('asymetric kernel', () => {
    const result = fftConvolution([1, 2, 3, 4], [-2, 0, 1]);
    expect(result).toStrictEqual([2, 1, 0, -6]);
  });

  it('cut border', () => {
    let result = fftConvolution([2, 4, 2], [-1, 0, 1], 'CUT');
    checkClose(result, [0]);

    result = fftConvolution([2, 4, 2], [2], 'CUT');
    expect(result).toStrictEqual([4, 8, 4]);

    result = fftConvolution(
      [0, 255, 255, 255, 255, 0, 0, 0],
      [0.1, 0.2, 0.3],
      'CUT'
    );
    checkClose(result, [127.5, 153, 153, 76.5, 25.5, 0]);
  });

  it('throws on invalid kernel', () => {
    expect(() => fftConvolution([1], [1, 1])).toThrow(
      /kernel must have an odd positive length. Got 2/
    );
  });

  it('throws on invalid border type', () => {
    expect(() => fftConvolution([1], [1], 'XXX')).toThrow(
      /unexpected border type: XXX/
    );
  });

  it('throws on invalid size', () => {
    expect(() => fftConvolution([], [1])).toThrow(
      /size must be a positive integer. Got 0/
    );
  });

  it('throws when input length does not match size', () => {
    const fft = new FFTConvolution(1, [1]);
    expect(() => fft.convolve([1, 2])).toThrow(
      /input length \(2\) does not match setup size \(1\)/
    );
  });
});

function checkClose(result, expected) {
  for (let i = 0; i < result.length; i++) {
    expect(result[i]).toBeCloseTo(expected[i], 12);
  }
}


src/__tests__/directConvolution.js
import { directConvolution, DirectConvolution } from '..';

describe('direct convolution', () => {
  it('with one value kernel', () => {
    expect(directConvolution([0, 1, 2, 3], [0])).toStrictEqual([0, 0, 0, 0]);
    expect(directConvolution([0, 1, 2, 3], [1])).toStrictEqual([0, 1, 2, 3]);
    expect(directConvolution([0, 1, 2, 3], [2])).toStrictEqual([0, 2, 4, 6]);
  });

  it('with three values kernel', () => {
    // prettier-ignore
    expect(directConvolution([0, 1, 2, 3], [1, 1, 1])).toStrictEqual([1, 3, 6, 5]);
    // prettier-ignore
    expect(directConvolution([0, 1, 2, 3], [-1, 1, -1])).toStrictEqual([-1, -1, -2, 1]);
  });

  it('asymetric kernel', () => {
    const result = directConvolution([1, 2, 3, 4], [-2, 0, 1]);
    expect(result).toStrictEqual([2, 1, 0, -6]);
  });

  it('cut border', () => {
    let result = directConvolution([2, 4, 2], [-1, 0, 1], 'CUT');
    expect(result).toStrictEqual([0]);

    result = directConvolution([2, 4, 2], [2], 'CUT');
    expect(result).toStrictEqual([4, 8, 4]);

    result = directConvolution(
      [0, 255, 255, 255, 255, 0, 0, 0],
      [0.1, 0.2, 0.3],
      'CUT'
    );
    expect(result).toStrictEqual([127.5, 153, 153, 76.5, 25.5, 0]);
  });

  it('throws on invalid kernel', () => {
    expect(() => directConvolution([1], [1, 1])).toThrow(
      /kernel must have an odd positive length. Got 2/
    );
  });

  it('throws on invalid border type', () => {
    expect(() => directConvolution([1], [1], 'XXX')).toThrow(
      /unexpected border type: XXX/
    );
  });

  it('throws on invalid size', () => {
    expect(() => directConvolution([], [1])).toThrow(
      /size must be a positive integer. Got 0/
    );
  });

  it('throws when input length does not match size', () => {
    const direct = new DirectConvolution(1, [1]);
    expect(() => direct.convolve([1, 2])).toThrow(
      /input length \(2\) does not match setup size \(1\)/
    );
  });
});


src/utils.js
export function checkSize(size) {
  if (!Number.isInteger(size) || size < 1) {
    throw new TypeError(`size must be a positive integer. Got ${size}`);
  }
}

export function checkKernel(kernel) {
  if (kernel.length === 0 || kernel.length % 2 !== 1) {
    throw new RangeError(
      `kernel must have an odd positive length. Got ${kernel.length}`
    );
  }
}

export function checkBorderType(borderType) {
  if (borderType !== 'CONSTANT' && borderType !== 'CUT') {
    throw new RangeError(`unexpected border type: ${borderType}`);
  }
}

export function checkInputLength(actual, expected) {
  if (actual !== expected) {
    throw new RangeError(
      `input length (${actual}) does not match setup size (${expected})`
    );
  }
}

export function createArray(len) {
  const array = [];
  for (var i = 0; i < len; i++) {
    array.push(0);
  }
  return array;
}


ml-convolution.d.ts
declare module 'ml-convolution' {
  export enum BorderType {
    CONSTANT = 'CONSTANT',
    CUT = 'CUT'
  }
  export function directConvolution(
    input: ArrayLike<number>,
    kernel: ArrayLike<number>,
    borderType?: BorderType
  ): number[];
  export class DirectConvolution {
    public constructor(
      size: number,
      kernel: ArrayLike<number>,
      borderType?: BorderType
    );
    public convolve(input: ArrayLike<number>): number[];
  }
  export function fftConvolution(
    input: ArrayLike<number>,
    kernel: ArrayLike<number>,
    borderType?: BorderType
  ): number[];
  export class FFTConvolution {
    public constructor(
      size: number,
      kernel: ArrayLike<number>,
      borderType?: BorderType
    );
    public convolve(input: ArrayLike<number>): number[];
  }
}


src/fftConvolution.js
import FFT from 'fft.js';
import nextPOT from 'next-power-of-two';

import {
  checkKernel,
  checkSize,
  checkBorderType,
  checkInputLength,
  createArray
} from './utils';

export class FFTConvolution {
  constructor(size, kernel, borderType = 'CONSTANT') {
    checkSize(size);
    checkKernel(kernel);
    checkBorderType(borderType);

    this.size = size;
    this.kernelOffset = (kernel.length - 1) / 2;
    this.doubleOffset = 2 * this.kernelOffset;
    this.borderType = borderType;
    const resultLength = size + this.doubleOffset;
    this.fftLength = nextPOT(Math.max(resultLength, 2));
    this.fftComplexLength = this.fftLength * 2;
    this.fft = new FFT(this.fftLength);

    kernel = kernel.slice().reverse();
    const paddedKernel = createArray(this.fftComplexLength);
    this.fftKernel = createArray(this.fftComplexLength);
    pad(kernel, paddedKernel, this.fftComplexLength);
    this.fft.transform(this.fftKernel, paddedKernel);

    this.paddedInput = createArray(this.fftComplexLength);
    this.fftInput = createArray(this.fftComplexLength);

    this.ifftOutput = createArray(this.fftComplexLength);
    this.result = paddedKernel;
  }

  convolve(input) {
    checkInputLength(input.length, this.size);
    pad(input, this.paddedInput, this.fftComplexLength);
    this.fft.transform(this.fftInput, this.paddedInput);

    for (var i = 0; i < this.fftInput.length; i += 2) {
      const tmp =
        this.fftInput[i] * this.fftKernel[i] -
        this.fftInput[i + 1] * this.fftKernel[i + 1];
      this.fftInput[i + 1] =
        this.fftInput[i] * this.fftKernel[i + 1] +
        this.fftInput[i + 1] * this.fftKernel[i];
      this.fftInput[i] = tmp;
    }

    this.fft.inverseTransform(this.ifftOutput, this.fftInput);
    const r = this.fft.fromComplexArray(this.ifftOutput, this.result);
    if (this.borderType === 'CONSTANT') {
      return r.slice(this.kernelOffset, this.kernelOffset + input.length);
    } else {
      return r.slice(this.doubleOffset, input.length);
    }
  }
}

export function fftConvolution(input, kernel, borderType) {
  return new FFTConvolution(input.length, kernel, borderType).convolve(input);
}

function pad(data, out, len) {
  let i = 0;
  for (; i < data.length; i++) {
    out[i * 2] = data[i];
    out[i * 2 + 1] = 0;
  }

  i *= 2;
  for (; i < len; i += 2) {
    out[i] = 0;
    out[i + 1] = 0;
  }
}


src/directConvolution.js
import {
  checkSize,
  checkKernel,
  checkBorderType,
  checkInputLength,
  createArray
} from './utils';

export class DirectConvolution {
  constructor(size, kernel, borderType = 'CONSTANT') {
    checkSize(size);
    checkKernel(kernel);
    checkBorderType(borderType);

    this.size = size;
    this.kernelOffset = (kernel.length - 1) / 2;
    this.outputSize =
      borderType === 'CONSTANT' ? size : size - 2 * this.kernelOffset;
    this.output = createArray(this.outputSize);
    this.kernel = kernel;
    this.kernelSize = kernel.length;
    this.borderType = borderType;
  }

  convolve(input) {
    checkInputLength(input.length, this.size);
    this.output.fill(0);
    if (this.borderType === 'CONSTANT') {
      this._convolutionBorder0(input);
    } else {
      this._convolutionBorderCut(input);
    }
    return this.output;
  }

  _convolutionBorder0(input) {
    for (let i = 0; i < this.size; i++) {
      for (let j = 0; j < this.kernelSize; j++) {
        this.output[i] +=
          interpolateInput(input, i - this.kernelOffset + j) * this.kernel[j];
      }
    }
  }

  _convolutionBorderCut(input) {
    for (let i = this.kernelOffset; i < this.size - this.kernelOffset; i++) {
      const index = i - this.kernelOffset;
      for (let j = 0; j < this.kernelSize; j++) {
        this.output[index] += input[index + j] * this.kernel[j];
      }
    }
  }
}

export function directConvolution(input, kernel, borderType) {
  return new DirectConvolution(input.length, kernel, borderType).convolve(
    input
  );
}

function interpolateInput(input, idx) {
  if (idx < 0) return 0;
  else if (idx >= input.length) return 0;
  return input[idx];
}


benchmark/convolution.js
'use strict';

const { FFTConvolution, DirectConvolution } = require('..');

const tests = {
  data: [128, 512, 2048, 4096, 16384, 65536, 262144, 1048576],
  kernel: [5, 11, 17, 33, 65, 129, 513]
};

function createArray(length) {
  const array = [];
  for (let i = 0; i < length; i++) {
    array.push(Math.random());
  }
  return array;
}

function test(dataLength, kernelLength) {
  const data = createArray(dataLength);
  const kernel = createArray(kernelLength);

  const direct = new DirectConvolution(dataLength, kernel);
  const fft = new FFTConvolution(dataLength, kernel);

  const fftResult = measure(data, fft);
  const directResult = measure(data, direct);

  return { fftResult, directResult };
}

function measure(data, convolution) {
  const start = Date.now();
  let shouldStop = false;
  let iterations = 0;
  while (!shouldStop) {
    convolution.convolve(data);
    iterations++;
    shouldStop = iterations >= 10 && Date.now() - start >= 1000;
  }
  const time = Date.now() - start;
  return {
    time,
    iterations
  };
}

console.log('| Data x Kernel | fft [ops/s] | direct [ops/s] |');
console.log('| ------------- | --- | ------ |');
for (const dataLength of tests.data) {
  for (const kernelLength of tests.kernel) {
    const result = test(dataLength, kernelLength);
    console.log(
      `| ${dataLength} x ${kernelLength} | ${formatResult(
        result.fftResult
      )} | ${formatResult(result.directResult)} |`
    );
  }
}

function formatResult(result) {
  return Math.round(result.iterations / (result.time / 1000));
}
